modules/assim.sequential/R/Multi_Site_Constructors.R

```diff
@@ -0,0 +1,258 @@
+  Y<-c()
+  
+  for (site in site.ids){
+    choose <- sapply(var.names, agrep, x=names(obs.t.mean[[site]]), max=1, USE.NAMES = F) %>% unlist
+    # if there is no obs for this site
+    if(length(choose)==0){
+      next;
+    }else{
+      Y <- c(Y, unlist(obs.t.mean[[site]][choose]))
+      # collecting them
+      site.specific.Rs <- c(site.specific.Rs, list(as.matrix(obs.t.cov[[site]][choose,choose])) )
+    }
+  #make block matrix out of our collection
+  R <- Matrix::bdiag(site.specific.Rs) %>% as.matrix()
+    }
+
+  return(list(Y=Y, R=R))
+}
+
+
+##' @title block_matrix
+##' @name  block_matrix
+##' @author Guy J. Abel
+##' 
+##' @param x Vector of numbers to identify each block.
+##' @param b Numeric value for the size of the blocks within the matrix ordered depending on byrow
+##' @param byrow ogical value. If FALSE (the default) the blocks are filled by columns, otherwise the blocks in the matrix are filled by rows.
+##' @param dimnames Character string of name attribute for the basis of the blcok matrix. If NULL a vector of the same length of b provides the basis of row and column names.#'.
+##’  
+##' 
+##' @description This function is adopted from migest package.
+##' 
+##' @return Returns a matrix with block sizes determined by the b argument. Each block is filled with the same value taken from x.
+##' @export
+block_matrix <- function (x = NULL, b = NULL, byrow = FALSE, dimnames = NULL) {
+  n <- length(b)
+  bb <- rep(1:n, times = b)
+  dn <- NULL
+  if (is.null(dimnames)) {
+    dn <- rep(1:n, times = b)
+    dd <- unlist(sapply(b, seq, from = 1))
+    dn <- paste0(dn, dd)
+    dn <- list(dn, dn)
+  }
+  if (!is.null(dimnames)) {
+    dn <- dimnames
+  }
+  xx <- matrix(NA, nrow = sum(b), ncol = sum(b), dimnames = dn)
+  k <- 1
+  if (byrow == TRUE) {
+    for (i in 1:n) {
+      for (j in 1:n) {
+        xx[i == bb, j == bb] <- x[k]
+        k <- k + 1
+      }
+    }
+  }
+  if (byrow == FALSE) {
+    for (j in 1:n) {
+      for (i in 1:n) {
+        xx[i == bb, j == bb] <- x[k]
+        k <- k + 1
+      }
+    }
+  }
+  return(xx)
+}
+
+##' @title Construct.H.multisite
+##' @name  Construct.H.multisite
+##' @author Hamze
+##' 
+##' @param site.ids a vector name of site ids  
+##' @param var.names vector names of state variable names
+##' @param obs.t.mean list of vector of means for the time t for different sites. 
+##' 
+##' @description This function is makes the blocked mapping function.
+##' 
+##' @return Returns a matrix with block sizes determined by the b argument. Each block is filled with the same value taken from x.
+##' @export
+Construct.H.multisite <- function(site.ids, var.names, obs.t.mean){
+
+  site.ids.with.data <- names(obs.t.mean)
+  site.specific.Hs <- list()
+  
+  nobs <- obs.t.mean %>% map_dbl(~length(.x)) %>% max
+  nsite <- length(site.ids)
+  nsite.ids.with.data <-length(site.ids.with.data)
+  nvariable <- length(var.names)
+  H <- matrix(0, (nobs * nsite.ids.with.data), (nvariable*nsite))
```

Comment:
This process of setting nobs to the maximum number of observations seen at any one site seems like it could result in H having rows that are all zeros, which doesn't seem desirable. Why not just sum over the number of observations?